apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Trace/TraceTest.cs

```diff
@@ -0,0 +1,420 @@
+    {
+        public override double GetSampleRate() => 1000000;
+        public override BufferOptions GetBufferOptions() =>
+            BufferOptions.SizedBuffer(1000);
+    }
+
+    /// <summary>
+    /// A web application to test <see cref="CloudTraceMiddleware"/> and associated classes.
+    /// This app does not use a buffer and will sample 1 QPS.
+    /// </summary>
+    public class TraceTestNoBufferApplication : AbstractTraceTestApplication
+    {
+        public override double GetSampleRate() => 1;
+        public override BufferOptions GetBufferOptions() => BufferOptions.NoBuffer();
+    }
+
+    /// <summary>
+    /// A web application to test <see cref="CloudTraceMiddleware"/> and associated classes.
+    /// This app does not use a buffer and will sample 1,000,000 QPS.
+    /// This will allow all calls to be traced and push them to the Trace API immediately.
+    /// </summary>
+    public class TraceTestNoBufferHighQpsApplication : AbstractTraceTestApplication
+    {
+        public override double GetSampleRate() => 1000000;
+        public override BufferOptions GetBufferOptions() => BufferOptions.NoBuffer();
+    }
+
+    /// <summary>
+    /// A web application to test <see cref="CloudTraceMiddleware"/> and associated classes.
+    /// This app does not use a buffer and will sample 0.0000001 QPS.
+    /// This will allow no calls to be traced (unless they have a trace header) and
+    /// push them to the Trace API immediately.
+    /// </summary>
+    public class TraceTestNoBufferLowQpsApplication : AbstractTraceTestApplication
+    {
+        public override double GetSampleRate() => 0.0000001;
+        public override BufferOptions GetBufferOptions() => BufferOptions.NoBuffer();
+    }
+
+    /// <summary>
+    /// A base web application to test <see cref="CloudTraceMiddleware"/> and associated classes.
+    /// </summary>
+    public abstract class AbstractTraceTestApplication
+    {
+        private readonly string _projectId;
+
+        public AbstractTraceTestApplication()
+        {
+            _projectId = Utils.GetProjectIdFromEnvironment();
+        }
+
+        public abstract double GetSampleRate();
+
+        public abstract BufferOptions GetBufferOptions();
+
+        public void ConfigureServices(IServiceCollection services)
+        {
+            var traceConfig = TraceConfiguration.Create(GetSampleRate(), GetBufferOptions());
+            services.AddGoogleTrace(_projectId, traceConfig);
+            services.AddMvc();
+        }
+
+        public void Configure(IApplicationBuilder app)
+        {
+            app.UseGoogleTrace();
+            app.UseMvc(routes =>
+            {
+                routes.MapRoute(
+                    name: "default",
+                    template: "{controller=Trace}/{action=Index}/{id}");
+            });
+        }
+    }
+
+    /// <summary>
+    /// A controller for the <see cref="TraceTestApplication"/> used trace calls.
```

Comment:
There doesn't seem to be a class named `TraceTestApplication`. Was this meant to be `AbstractTraceTestApplication`?